service/installer/helm/helm.go

```diff
@@ -0,0 +1,355 @@
+
+		// Settings.
+		helmBinaryPath: config.HelmBinaryPath,
+		organisation:   config.Organisation,
+		password:       config.Password,
+		registry:       config.Registry,
+		username:       config.Username,
+	}
+
+	if err := installer.login(); err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return installer, nil
+}
+
+func (i *Installer) Install(project spec.Project) error {
+	i.logger.Log("debug", "ensuring chart is installed", "name", project.Name, "ref", project.Ref)
+
+	var err error
+
+	// We create a tmp dir in which all Helm values files and tarballs are written
+	// to. After we are done we can just remove the whole tmp dir to clean up.
+	var tmpDir string
+	{
+		tmpDir, err = afero.TempDir(i.fileSystem, "", "draughtsman-operator-helm-installer")
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		defer func() {
+			err := i.fileSystem.RemoveAll(tmpDir)
+			if err != nil {
+				i.logger.Log("error", fmt.Sprintf("could not remove tmp dir: %#v", err), "name", project.Name, "ref", project.Ref)
+			}
+		}()
+	}
+
+	var tarballPath string
+	{
+		tarballPath = path.Join(tmpDir, i.tarballName(project))
+
+		_, err := i.runHelmCommand("pull", "registry", "pull", "--dest", tmpDir, "--tarball", i.versionedChartName(project))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		exists, err := afero.Exists(i.fileSystem, tarballPath)
+		if !exists {
+			return microerror.Maskf(helmError, "could not find downloaded tarball at '%s'", tarballPath)
+		}
+
+		i.logger.Log("debug", "downloaded chart", "tarball", tarballPath)
+	}
+
+	// The intaller accepts multiple configurers during initialization. Here we
+	// iterate over all of them to get all the values they provide. For each
+	// values file we have to create a file in the tmp dir we created above.
+	var valuesFilesArgs []string
+	for _, c := range i.configurers {
+		fileName := filepath.Join(tmpDir, fmt.Sprintf("%s-values.yaml", strings.ToLower(string(c.Type()))))
+		values, err := c.Values()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = afero.WriteFile(i.fileSystem, fileName, []byte(values), os.FileMode(0644))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		valuesFilesArgs = append(valuesFilesArgs, "--values", fileName)
+	}
+
+	// The arguments used to execute Helm for app installation can take multiple
+	// values files. At the end the command looks something like this.
+	//
+	//     helm upgrade --install --values ${file1} --values $(file2) ${project} ${tarball_path}
+	//
+	var installCommand []string
+	{
+		installCommand = append(installCommand, "upgrade", "--install")
+		installCommand = append(installCommand, valuesFilesArgs...)
+		installCommand = append(installCommand, project.Name, tarballPath)
+
+		_, err := i.runHelmCommand("install", installCommand...)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (i *Installer) List(projects []spec.Project) ([]spec.Project, error) {
```

Comment:
The list functionality is new because the project resource needs to know the current state to be able to reconcile.